### PR TITLE
feat: Add verify-token Action

### PR DESCRIPTION
**Pull request by laurentsimon**
Signed-off-by: laurentsimon <laurentsimon@google.com>

### CODE REVIEW

**Comment by laurentsimon** [comment]
note: setup-token pre-submit is being fixed in https://github.com/slsa-framework/slsa-github-generator/pull/1417 and should be merged before this PR

File: .github/actions/verify-token/src/main.ts [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    // Verify the intended recipient.
+    validateField(
+      "builder.audience",
+      rawTokenObj.builder.audience,
+      workflowRecipient
+    );
+
+    // Verify that the runner label is not empty.
+    validateNonEmptyField(
+      "builder.runner_label",
```
Comment (laurentsimon) [inline .github/actions/verify-token/src/main.ts:120]: we could also verify that it's one of the supported labels. Currently we only support `ubuntu-latest`. Shall we do a verification here?

File: .github/actions/verify-token/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+      throw new Error(`malformed token: ${parts.length}`);
+    }
+
+    const bundleStr = Buffer.from(parts[0], "base64").toString("utf-8");
+    const b64Token = parts[1];
+    const bundle = JSON.parse(bundleStr);
+
+    // First, verify the signature, i.e., that it is signed by a certificate that
+    // chains up to Fulcio.
+    await sigstore.sigstore.verify(bundle, Buffer.from(b64Token));
```
Comment (asraa) [inline .github/actions/verify-token/src/main.ts:74]: so the payload signed is the base64-encoded token? why not the rawToken?

File: .github/actions/verify-token/src/main.ts (hunk 1 above)
Comment (asraa) [reply .github/actions/verify-token/src/main.ts:120]: sgtm

**Review by asraa — APPROVED** [review]
nice!

File: .github/actions/verify-token/src/main.ts (hunk 1 above)
Comment (laurentsimon) [reply .github/actions/verify-token/src/main.ts:120]: done

File: .github/actions/verify-token/src/main.ts (hunk 2 above)
Comment (laurentsimon) [reply .github/actions/verify-token/src/main.ts:74]: 1. less parsing before verification, so less attack surface
2. base64 decoding allows some bits to be flipped without changing the decoded results, which seems weird for signature verification to still pass

Not a big deal, I agree :)

**Pull request merged**